### PR TITLE
`reserveCells` crashes `mint` mesh inside Umpire

### PR DESCRIPTION
**Pull request by gunney1**
# Crash reserving space for `UnstructuredMesh`

- This PR is a bug fix
- It does the following:
  - Adds tests that verify the crash (which happens on our docker platforms)
  - Fix the bug.

The cause of the bug is the combination of static Umpire library with shared Axom library in Docker.  Somehow, this resulted in 2 singleton `ResourceManagers`.  The root cause of the crash in `reserveCells` was when the second singleton tried to reallocate memory allocated by the first.  Fixed by building Umpire shared.  We updated the Docker image to make Umpire shared.

This crash had been plaguing PR https://github.com/LLNL/axom/pull/1263.  The `MarchingCubes` test crashed while asking `mint::UnstructuredMesh` to reserve space.

Two identical tests were added to monitor this behavior:
- In `src/axom/quest/tests/quest_initialize.cpp`
- In `src/axom/mint/tests/mint_mesh.cpp`

Interesting behaviors:
- The test added to `quest` reproduced the crash, but the one added to `mint` did not.
- The crashes only occurred in our docker builds.

Addresses issue https://github.com/LLNL/axom/issues/1287

### CODE REVIEW

**Comment by gunney1** [comment]
Some new observations:

1. Based on @white238's guess, I disabled MPI (and MFEM, which was configured with MPI) from the failing config.  This fixed the failure.  Not exactly sure what the mechanism of failure was.
2. I tried to construct the `UnstructuredMesh` without using Sidre.  This also gets past the crash.  This provides a work-around for this bug, but it effectively disables some Sidre-dependent features when using the `UnstructuredMesh`.

**Comment by gunney1** [comment]
After merging in the latest develop branch, the failed test (`quest_initialize`) continues to fail.  But if I remove the `-DBUILD_SHARED_LIBS=ON` from the configuration, the tests pass.  @white238 What do you think of this clue?

File: azure-pipelines.yml [hunk 1]
```diff
@@ -9,8 +9,8 @@ variables:
   DO_BUILD: 'yes'
   DO_TEST: 'yes'
   DO_CLEAN: 'no'
-  CLANG14_IMAGENAME: 'axom/tpls:clang-14_09-23-24_17h-07m'
-  GCC13_IMAGENAME: 'axom/tpls:gcc-13_09-23-24_17h-07m'
+  CLANG14_IMAGENAME: 'axom/tpls:clang-14_10-10-24_01h-31m'
```
Comment (rhornung67) [inline azure-pipelines.yml:12]: Is this change required?

Comment (gunney1) [reply azure-pipelines.yml:12]: Yes.  The new image has Umpire shared libs.  That's the key to the bug fix.

**Review by rhornung67 — APPROVED** [review]

File: azure-pipelines.yml (hunk 1 above)
Comment (rhornung67) [reply azure-pipelines.yml:12]: @white238 @kennyweiss are you OK with switching to Umpire shared libs for azure testing?

Comment (white238) [reply azure-pipelines.yml:12]: We (@kennyweiss , @gunney1 , @davidbeckingsale ) met about this last night. Basically a static umpire with a shared axom creates multiple singletons in Umpire. We need to not allow this to happen in our Spack recipe as well but this seems to be the only way around this.

Comment (rhornung67) [reply azure-pipelines.yml:12]: Do you mean add a constraint in the Axom Spack package that static Umpire conflicts with shared Axom?

Comment (white238) [reply azure-pipelines.yml:12]: yes and/or a conflict

Comment (kennyweiss) [reply azure-pipelines.yml:12]: Yes -- the three of us had a meeting yesterday w/ @davidbeckingsale and concluded that the static umpire combined w/ a shared axom in the docker configuration was leading to more than one umpire instance. 

There might be a solution to this w/ improved linking orders and/or linker flags, but it will be difficult to track down in the short term, and is not necessary for @gunney1's actual deliverable (i.e. the CI is testing a different configuration than the one we actually need). @gunney1 is planning to create an issue summarizing that discussion so we can look into it in the future. 

It's also worth noting that we've encountered similar issues in the past when mixing libraries w/ singletons with upstream shared or Fortran libraries (e.g. #865 and #270).

File: scripts/spack/configs/docker/ubuntu22/spack.yaml [hunk 2]
```diff
@@ -165,7 +165,7 @@ spack:
     scr:
       require: "@3.0.1~shared~tests~examples"
     umpire:
-      require: "@2024.07.0~shared~examples~werror"
+      require: "@2024.07.0+shared~examples~werror"
```
Comment (kennyweiss) [inline scripts/spack/configs/docker/ubuntu22/spack.yaml:168]: @rhornung67 -- the change to the docker TPL configuration is actually in this line -- we are now forcing Umpire to build as a shared library for better compatibility w/ the shared axom config.

File: src/axom/quest/tests/quest_initialize.cpp [hunk 3]
```diff
@@ -72,6 +75,76 @@ TEST(quest_initialize, signed_distance_pointer_initialize)
   delete input_mesh;
 }
 
+// Test allocation/reallocation using axom::allocate and axom::reallocate bytes
```
Comment (kennyweiss) [inline src/axom/quest/tests/quest_initialize.cpp:78]: Thanks for adding these reproducers @gunney1 

Please add some comments to this file about why these tests are here, rather than in `sidre` tests of `core` tests, since it is not obvious.  Specifically, that these tests are reproducers for bugs that we found w/ a static config of umpire mixed with a shared config of axom, and that the failures only appear when there are several shared libs that include the static umpire, e.g. `axom::core`, `axom::sidre`, `axom::mint` and `axom::quest`).

File: src/axom/quest/tests/quest_initialize.cpp [hunk 4]
```diff
@@ -72,6 +75,76 @@ TEST(quest_initialize, signed_distance_pointer_initialize)
...
+TEST(quest_initialize, immediate_ug_reserve)
+{
+  axom::sidre::DataStore dataStore;
+  axom::sidre::Group* meshGroup = dataStore.getRoot()->createGroup("myGroup");
+  axom::mint::UnstructuredMesh<axom::mint::SINGLE_SHAPE> contourMesh(
+    2,
+    axom::mint::CellType::SEGMENT,
+    meshGroup);
+  contourMesh.reserveCells(10);  // This may unexpectedly crash.
+}
```
Comment (kennyweiss) [inline src/axom/quest/tests/quest_initialize.cpp:145]: I think this test needs to be guarded by `AXOM_MINT_USE_SIDRE`

File: src/axom/quest/tests/quest_initialize.cpp [hunk 5]
```diff
@@ -6,6 +6,9 @@
 // Axom includes
 #include "axom/mint.hpp"
```
Comment (kennyweiss) [inline src/axom/quest/tests/quest_initialize.cpp:7]: We should probably explicitly include `axom/config.hpp` here 
(even though it is indirectly included by `axom/mint.hpp`)

**Review by kennyweiss — APPROVED** [review]
Thanks @gunney1 

Please also test this branch w/ `AXOM_MINT_USE_SIDRE=OFF` to ensure that all sidre usage w/ mint is properly guarded (see: https://github.com/LLNL/axom/issues/1435).

File: src/axom/quest/tests/quest_initialize.cpp (hunk 3 above)
Comment (gunney1) [reply src/axom/quest/tests/quest_initialize.cpp:78]: Done.

File: src/axom/quest/tests/quest_initialize.cpp (hunk 4 above)
Comment (gunney1) [reply src/axom/quest/tests/quest_initialize.cpp:145]: Yes.  Fixed.

File: src/axom/quest/tests/quest_initialize.cpp (hunk 5 above)
Comment (gunney1) [reply src/axom/quest/tests/quest_initialize.cpp:7]: Done.
